Stage 2 clustering emulator

EmyrClement opened the pull request:
This is a port of the existing PR to the [PFCal-dev fork](https://github.com/PFCal-dev/cmssw/pull/499)

jbsauvan commented:
Some quick general comments:
- It might help the readability if the different functions could be grouped together in different higher-level classes. At the moment i'm afraid it could quickly become difficult to maintain when we want to try and compare some variations/updates of the different blocks.  For instance it could be split like:
  - TC distribution
  - seeding
  - clustering
  - cluster properties  
- The interfaces of the functions could be improved. They are not very explicit at the moment. For instance the function `clusterizer` has only in-out parameters (non-const references), so it is difficult to see what are the inputs and what are the outputs.  
- Magic numbers and naming could be improved.
  - In particular it is not very clear from their names what the different TC distribution functions are doing.
  - There are still many magic numbers. If these are not meant to be parameters, they should still be defined as named constant to understand what they are
- There is a massive usage of `shared_ptr`. 
  - In general `unique_ptr` should be preferred  as the ownership is clearer in that case and `shared_ptr` have an overhead. 
  - `shared_ptr` should only be used if there is a clear need of shared ownership. Is it the case?

EmyrClement commented:
Thanks for the comments.

> Some quick general comments:
> 
> * It might help the readability if the different functions could be grouped together in different higher-level classes. At the moment i'm afraid it could quickly become difficult to maintain when we want to try and compare some variations/updates of the different blocks.  For instance it could be split like:
>   
>   * TC distribution
>   * seeding
>   * clustering
>   * cluster properties
> * The interfaces of the functions could be improved. They are not very explicit at the moment. For instance the function `clusterizer` has only in-out parameters (non-const references), so it is difficult to see what are the inputs and what are the outputs.

I agree with both these points.  I think your suggestion on the splitting of the different code blocks make sense, so will start with that proposal.  I can also improve the interfaces of the functions.

> * Magic numbers and naming could be improved.
>   
>   * In particular it is not very clear from their names what the different TC distribution functions are doing.
>   * There are still many magic numbers. If these are not meant to be parameters, they should still be defined as named constant to understand what they are

Apologies, I hadn't yet moved all magic numbers into the python configuration.  For the TC distribution functions, whilst there may not be better names for them, I can at least add comments to explain what each one does as a first step.

I have one question on accessing the configuration parameters.  We set up the `HGCalHistoClusteringWrapper::configure` function to take the `edm::ParameterSet`, extract the configuration parameters for a particular FPGA/region, and store those parameters in a class.  As this method is called multiple times per event (every time just before we call `HGCalHistoClusteringWrapper::process`), and the configuration for a particular FPGA/region stays the same event-by-event, I guess it would be better to construct the configuration class for all FPGAs/regions once per CMSSW job and then the `configure` method chooses which configuration class to use, rather than accessing the `edm::ParameterSet` each time?

> * There is a massive usage of `shared_ptr`.
>   
>   * In general `unique_ptr` should be preferred  as the ownership is clearer in that case and `shared_ptr` have an overhead.
>   * `shared_ptr` should only be used if there is a clear need of shared ownership. Is it the case?

I originally used `shared_ptr` as during the TC distribution, the TCs are stored in two vectors (a 1D and 2D vector) in a few places e.g.:

https://github.com/EmyrClement/cmssw/blob/S2ClusteringEmulatorIntegration_hgctpg/L1Trigger/L1THGCal/src/backend_emulator/HGCalHistoClusteringImpl_SA.cc#L128-L129

I think this requires the use of a `shared_ptr` (or a copy of the object).  Though I think later on, once we get to `HGCalHistoClusteringImplSA::triggerCellToHistogramCell`, we should be able to use `unique_ptr` for the TCs from then on, and certainly for pointers to other objects (histogram cells, clusters).

jbsauvan commented:
> Apologies, I hadn't yet moved all magic numbers into the python configuration. 

I thinks it is not necessary to put all these numbers in a python config, only those that will actually need to be configured could go there. But for those not going in python configs they should be defined as named constants.

> For the TC distribution functions, whilst there may not be better names for them, I can at least add comments to explain what each one does as a first step.

Ok. Agreed.

jbsauvan commented:
> I have one question on accessing the configuration parameters. We set up the `HGCalHistoClusteringWrapper::configure` function to take the `edm::ParameterSet`, extract the configuration parameters for a particular FPGA/region, and store those parameters in a class. As this method is called multiple times per event (every time just before we call `HGCalHistoClusteringWrapper::process`), and the configuration for a particular FPGA/region stays the same event-by-event, I guess it would be better to construct the configuration class for all FPGAs/regions once per CMSSW job and then the `configure` method chooses which configuration class to use, rather than accessing the `edm::ParameterSet` each time?

Indeed that might be better, given that the number of different configurations / FPGAs is not huge

jbsauvan commented:
> I think this requires the use of a `shared_ptr` (or a copy of the object). Though I think later on, once we get to `HGCalHistoClusteringImplSA::triggerCellToHistogramCell`, we should be able to use `unique_ptr` for the TCs from then on, and certainly for pointers to other objects (histogram cells, clusters).

Ok, I see.
Still I think that `shared_ptr` should only be used when necessary, in a minimum number of places.

EmyrClement commented:
The latest commits address most of the comments that were made, in particular:
- Reduce usage of `shared_ptr` and use `unique_ptr` where possible.
- Improve interfaces to the functions.  It should be clearer now which collections are inputs and outputs (fewer in-out parameters).  I also added comments to describe the functionality of the less obvious parts of the algorithm, in particular the TCDistribution steps.  However, I have kept the function names the same, so they match the naming convention in the firmware and python emulator.  There are also some steps in the TCDistribution that I don't completely understand how they work, but the comments aim to explain what they are trying to achieve.
- The different parts of the algorithm (TCDistribution, seeding, clustering, properties) have been split up into separate classes.
- Almost all "magic numbers" are now read from the python config.  It wasn't clear to me precisely what some parameters where e.g. in a few places the latency of particular internal steps of the algorithm is required, so I will need checking with Andy what these numbers represent.

In addition to the addressing the comments:
- I added a simple (not firmware based) implementation of a local maxima finder, which we use to compare with the existing CMSSW simulation.
- Store the cluster constituents in the output of the emulated clusters.  These correspond to the original CMSSW trigger cells, but the properties of the output clusters correspond to those calculated by the emulation.  This should allow a comparison of cluster properties calculated from the CMSSW TCs that have been clustered by the emulator, and the cluster properties calculated directly by the emulator.
- @jonathon-langford also modifed the ntuple producer to store the cluster properties that are produced by the emulation.

jbsauvan commented:
Thanks @EmyrClement !
Could you run `scram b code-format` and `scram b code-checks` and apply the proposed changes?

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalClusterT.h:
```diff
@@ -179,6 +179,69 @@ namespace l1t {
     void triggerCells90percent(float triggerCells90percent) { triggerCells90percent_ = triggerCells90percent; }
 
+    // Firmware-specific cluster properties
+    unsigned long int Sigma_E_Quotient() const { return Sigma_E_Quotient_SA_; }
+    unsigned long int Sigma_E_Fraction() const { return Sigma_E_Fraction_SA_; }
+    unsigned long int Mean_z_Quotient() const { return Mean_z_Quotient_SA_; }
+    unsigned long int Mean_z_Fraction() const { return Mean_z_Fraction_SA_; }
+    unsigned long int Mean_phi_Quotient() const { return Mean_phi_Quotient_SA_; }
+    unsigned long int Mean_phi_Fraction() const { return Mean_phi_Fraction_SA_; }
+    unsigned long int Mean_eta_Quotient() const { return Mean_eta_Quotient_SA_; }
+    unsigned long int Mean_eta_Fraction() const { return Mean_eta_Fraction_SA_; }
+    unsigned long int Mean_roz_Quotient() const { return Mean_roz_Quotient_SA_; }
+    unsigned long int Mean_roz_Fraction() const { return Mean_roz_Fraction_SA_; }
+    unsigned long int Sigma_z_Quotient() const { return Sigma_z_Quotient_SA_; }
+    unsigned long int Sigma_z_Fraction() const { return Sigma_z_Fraction_SA_; }
+    unsigned long int Sigma_phi_Quotient() const { return Sigma_phi_Quotient_SA_; }
+    unsigned long int Sigma_phi_Fraction() const { return Sigma_phi_Fraction_SA_; }
+    unsigned long int Sigma_eta_Quotient() const { return Sigma_eta_Quotient_SA_; }
+    unsigned long int Sigma_eta_Fraction() const { return Sigma_eta_Fraction_SA_; }
+    unsigned long int Sigma_roz_Quotient() const { return Sigma_roz_Quotient_SA_; }
+    unsigned long int Sigma_roz_Fraction() const { return Sigma_roz_Fraction_SA_; }
+    unsigned long int E_EM_over_E_Quotient() const { return E_EM_over_E_Quotient_SA_; }
+    unsigned long int E_EM_over_E_Fraction() const { return E_EM_over_E_Fraction_SA_; }
+    unsigned long int E_EM_core_over_E_EM_Quotient() const { return E_EM_core_over_E_EM_Quotient_SA_; }
+    unsigned long int E_EM_core_over_E_EM_Fraction() const { return E_EM_core_over_E_EM_Fraction_SA_; }
+    unsigned long int E_H_early_over_E_Quotient() const { return E_H_early_over_E_Quotient_SA_; }
+    unsigned long int E_H_early_over_E_Fraction() const { return E_H_early_over_E_Fraction_SA_; }
```
- Would be better to start method names (as well as attribute names below) with non-capital letters, as in principle capital letters are reserved for class names.
- Also would be good to prepend the name with a pattern that indicates that these are hardware variables (typically `hw` is used)

EmyrClement replied:
Could you clarify - do you mean the method names could be prepended with `hw`, or the attribute names?  Or both?

jbsauvan replied:
Both would be better I think, in order to keep the same names for the attributes and the getters

EmyrClement replied:
I may also remove the `_SA` prefix of the attribute names, as I think that was mainly from the first tests of the "standalone" framework.

jbsauvan commented on L1Trigger/L1THGCal/interface/backend_emulator/CentroidHelper.h:
```diff
@@ -0,0 +1,67 @@
+    unsigned int row_;
+    unsigned int energy_;
+    unsigned int X_;
+    unsigned int Y_;
```
Better to avoid starting attribute name with capital letters. Reserved for class names.

jbsauvan commented on L1Trigger/L1THGCal/interface/backend_emulator/DistServer.h:
```diff
@@ -0,0 +1,33 @@
+    DistServer(unsigned int nInputs, unsigned int nOutputs, unsigned int nInterleaving);
+    ~DistServer() {}
+
+    HGCalTriggerCellSAShrPtrCollection clock(HGCalTriggerCellSAShrPtrCollection& inputs);
```
In the `.cc` file the parameter is named differently: `data` instead of `inputs`
The variable name is not needed here in the `.h` so I would suggest just omitting the variable name here.

jbsauvan commented on L1Trigger/L1THGCal/interface/backend_emulator/DistServer.h:
```diff
@@ -0,0 +1,33 @@
+    unsigned int nInputs() const { return nInputs_; }
+    unsigned int nOutputs() const { return nOutputs_; }
+    unsigned int nInterleaving() const { return nInterleaving_; }
+    std::vector<std::vector<unsigned int> >& addr() { return addr_; }
```
Do we want to return a non-const reference? Meaning the private member `addr_` can be modified outside the class?

EmyrClement replied:
Made method private, only used by class internally.

jbsauvan commented on L1Trigger/L1THGCal/interface/backend_emulator/HGCalCluster_SA.h:
```diff
@@ -0,0 +1,274 @@
+                                                           unsigned long int Sum_Wc2,
+                                                           unsigned int Sum_Wc);
+    std::pair<unsigned int, unsigned int> Energy_ratio(unsigned int E_N, unsigned int E_D);
+    std::vector<int> ShowerLengthProperties(unsigned long int layerBits);
```
Same comment as before regarding capitalized names.

jbsauvan commented on L1Trigger/L1THGCal/interface/backend_emulator/HGCalCluster_SA.h:
```diff
@@ -0,0 +1,274 @@
+    unsigned long int E_EM_core_over_E_EM_Quotient() const { return E_EM_core_over_E_EM_Quotient_; }
+    unsigned long int E_EM_core_over_E_EM_Fraction() const { return E_EM_core_over_E_EM_Fraction_; }
+    unsigned long int E_H_early_over_E_Quotient() const { return E_H_early_over_E_Quotient_; }
+    unsigned long int E_H_early_over_E_Fraction() const { return E_H_early_over_E_Fraction_; }
```
Same comment as before regarding capitalized names.

jbsauvan commented on L1Trigger/L1THGCal/interface/backend_emulator/CentroidHelper.h:
```diff
@@ -0,0 +1,67 @@
+    void setIndex(const unsigned int index) { index_ = index; }
+
+    // Operators
+    bool operator>(const CentroidHelper& ch_other) { return true; }  // Nothing implemented?
```
Is it used somewhere?

jbsauvan commented on L1Trigger/L1THGCal/interface/backend_emulator/HGCalHistoClusteringConfig_SA.h:
```diff
@@ -0,0 +1,302 @@
+    void setZSide(const int zSide) { zSide_ = zSide; }
+    int zSide() const { return zSide_; }
+
+    void setStepLatencies(const std::vector<unsigned int> latencies);
```
Should rather pass `latencies` as const reference

EmyrClement replied:
Done

jbsauvan commented on L1Trigger/L1THGCal/interface/backend_emulator/HGCalHistoClusteringConfig_SA.h:
```diff
@@ -0,0 +1,302 @@
+
+  enum Step {
+    Uninitialized = -1,
+    Input = 0,
+    Dist0 = 1,
+    Dist1 = 2,
+    Dist2 = 3,
+    Dist3 = 4,
+    Dist4 = 5,
+    Dist5 = 6,
+    TcToHc = 7,
+    Hist = 8,
+    Smearing1D = 9,
+    NormArea = 10,
+    Smearing2D = 11,
+    Maxima1D = 12,  // Not actually used currently
+    Maxima2D = 13,
+    CalcAverage = 14,
+    Clusterizer = 15,
+    TriggerCellToCluster = 16,
+    ClusterSum = 17
```
It's actually safer not to specify enum values if possible to avoid typing errors and ensure that values are consecutive.

EmyrClement replied:
Enum values removed.

jbsauvan commented on L1Trigger/L1THGCal/interface/backend_emulator/HGCalHistoClusteringConfig_SA.h:
```diff
@@ -0,0 +1,302 @@
+
+    unsigned int cosLUT(unsigned int iBin) const { return cosLUT_.at(iBin); }
+
+    void setDepths(const std::vector<unsigned int> depths) {
```
Use const reference to pass `depths`

jbsauvan commented on L1Trigger/L1THGCal/interface/backend_emulator/HGCalHistoClusteringConfig_SA.h:
```diff
@@ -0,0 +1,302 @@
+
+    void setDepths(const std::vector<unsigned int> depths) {
+      depths_.clear();
+      for (const auto& depth : depths)
```
The size of `depths` is known so could reserve the size.

EmyrClement replied:
Done

jbsauvan commented on L1Trigger/L1THGCal/interface/backend_emulator/HGCalHistoClusteringConfig_SA.h:
```diff
@@ -0,0 +1,302 @@
+      for (const auto& depth : depths)
+        depths_.push_back(depth);
+    }
+    std::vector<unsigned int> depths() const { return depths_; }
```
Would be better to return a const ref to avoid the copy

jbsauvan commented on L1Trigger/L1THGCal/interface/backend_emulator/HGCalHistoClusteringConfig_SA.h:
```diff
@@ -0,0 +1,302 @@
+      for (const auto& triggerLayer : triggerLayers)
+        triggerLayers_.push_back(triggerLayer);
+    }
+    std::vector<unsigned int> triggerLayers() const { return triggerLayers_; }
```
Same comments as before

jbsauvan commented on L1Trigger/L1THGCal/interface/backend_emulator/HGCalHistoClusteringConfig_SA.h:
```diff
@@ -0,0 +1,302 @@
+      for (const auto& weight : layerWeights_E)
+        layerWeights_E_.push_back(weight);
+    }
+    std::vector<unsigned int> layerWeights_E() const { return layerWeights_E_; }
```
Same comments as before

jbsauvan commented on L1Trigger/L1THGCal/interface/backend_emulator/HGCalHistoClusteringConfig_SA.h:
```diff
@@ -0,0 +1,302 @@
+      for (const auto& weight : layerWeights_E_EM)
+        layerWeights_E_EM_.push_back(weight);
+    }
+    std::vector<unsigned int> layerWeights_E_EM() const { return layerWeights_E_EM_; }
```
Same comments as before

jbsauvan commented on L1Trigger/L1THGCal/interface/backend_emulator/HGCalHistoClusteringConfig_SA.h:
```diff
@@ -0,0 +1,302 @@
+      for (const auto& weight : layerWeights_E_EM_core)
+        layerWeights_E_EM_core_.push_back(weight);
+    }
+    std::vector<unsigned int> layerWeights_E_EM_core() const { return layerWeights_E_EM_core_; }
```
Same comments as before

jbsauvan commented on L1Trigger/L1THGCal/interface/backend_emulator/HGCalHistoClusteringConfig_SA.h:
```diff
@@ -0,0 +1,302 @@
+      for (const auto& weight : layerWeights_E_H_early)
+        layerWeights_E_H_early_.push_back(weight);
+    }
+    std::vector<unsigned int> layerWeights_E_H_early() const { return layerWeights_E_H_early_; }
```
Same comments as before

jbsauvan commented on L1Trigger/L1THGCal/interface/backend_emulator/HGCalHistoClusteringConfig_SA.h:
```diff
@@ -0,0 +1,302 @@
+    unsigned int nRowsForClustering_;
+    unsigned int clusterizerMagicTime_;
+
+    std::map<Step, unsigned int> stepLatency_;
```
Is a map needed? Steps are contiguous integer, so I guess a vector would make it?

EmyrClement replied:
Yes, a vector does work.  Changed it from map to vector.

jbsauvan commented on L1Trigger/L1THGCal/interface/backend_emulator/HGCalHistoClusteringImpl_SA.h:
```diff
@@ -0,0 +1,37 @@
+
+  class HGCalHistoClusteringImplSA {
+  public:
+    HGCalHistoClusteringImplSA(l1thgcfirmware::ClusterAlgoConfig& config);
```
Const reference would be better

EmyrClement replied:
Done

jbsauvan commented on L1Trigger/L1THGCal/interface/backend_emulator/HGCalHistoClusteringImpl_SA.h:
```diff
@@ -0,0 +1,37 @@
+                      HGCalClusterSAPtrCollection& clusterSums) const;
+
+  private:
+    l1thgcfirmware::ClusterAlgoConfig& config_;
```
Should it be a const reference?

EmyrClement replied:
Yes, done.

jbsauvan commented on L1Trigger/L1THGCal/interface/backend_emulator/HGCalHistoClustering_SA.h:
```diff
@@ -0,0 +1,37 @@
+
+  class HGCalHistoClustering {
+  public:
+    HGCalHistoClustering(l1thgcfirmware::ClusterAlgoConfig& config);
```
Const reference would be better

EmyrClement replied:
Done

jbsauvan commented on L1Trigger/L1THGCal/interface/backend_emulator/HGCalHistoClustering_SA.h:
```diff
@@ -0,0 +1,37 @@
+    void triggerCellToCluster(const l1thgcfirmware::HGCalTriggerCellSAShrPtrCollection& clusteredTriggerCells,
+                              l1thgcfirmware::HGCalClusterSAPtrCollection& clustersOut) const;
+
+    l1thgcfirmware::ClusterAlgoConfig& config_;
```
Same as before: const reference?

EmyrClement replied:
Done, same for next few similar comments.

jbsauvan commented on L1Trigger/L1THGCal/interface/backend_emulator/HGCalHistoSeeding_SA.h:
```diff
@@ -0,0 +1,42 @@
+
+  class HGCalHistoSeeding {
+  public:
+    HGCalHistoSeeding(l1thgcfirmware::ClusterAlgoConfig& config);
```
const reference?

jbsauvan commented on L1Trigger/L1THGCal/interface/backend_emulator/HGCalHistoSeeding_SA.h:
```diff
@@ -0,0 +1,42 @@
+    // For debugging
+    void printHistogram(const l1thgcfirmware::HGCalHistogramCellSAPtrCollection& histogram) const;
+
+    l1thgcfirmware::ClusterAlgoConfig& config_;
```
const reference?

jbsauvan commented on L1Trigger/L1THGCal/interface/backend_emulator/HGCalHistogramCell_SA.h:
```diff
@@ -0,0 +1,78 @@
+                       unsigned int sortKey)
+        : clock_(clock),
+          index_(index),
+          S_(S),
+          X_(X),
+          Y_(Y),
+          N_(N),
```
X and Y  I can guess what it is. But S and N?
It would be good to find better names if possible.

EmyrClement replied:
S is the sum of the bin content, and N is the number of cells in a bin.  I would prefer to keep this naming to match the firmware and python emulator.  I can add comments explaining their meanings.

jbsauvan replied:
I would say then the firmware naming should be changed :grin: 
But ok to add comments and keep the names in sync.

jbsauvan commented on L1Trigger/L1THGCal/interface/backend_emulator/HGCalHistogramCell_SA.h:
```diff
@@ -0,0 +1,78 @@
+        : clock_(0), index_(0), S_(0), X_(0), Y_(0), N_(0), sortKey_(0), frameValid_(false), dataValid_(false) {}
+
+    HGCalHistogramCell(unsigned int clock, unsigned int index, unsigned int sortKey)
+        : HGCalHistogramCell(4 * sortKey + clock, index, 0, 0, 0, 0, sortKey) {}
```
Magic number `4`

jbsauvan commented on L1Trigger/L1THGCal/interface/backend_emulator/HGCalTCDistribution_SA.h:
```diff
@@ -0,0 +1,62 @@
+
+  class HGCalTCDistribution {
+  public:
+    HGCalTCDistribution(l1thgcfirmware::ClusterAlgoConfig& config);
```
const reference?

jbsauvan commented on L1Trigger/L1THGCal/interface/backend_emulator/HGCalTCDistribution_SA.h:
```diff
@@ -0,0 +1,62 @@
+                        unsigned int nInterleave,
+                        bool setOutputGrid) const;
+
+    l1thgcfirmware::ClusterAlgoConfig& config_;
```
const reference?

jbsauvan commented on L1Trigger/L1THGCal/interface/backend_emulator/HGCalTriggerCell_SA.h:
```diff
@@ -0,0 +1,89 @@
+    unsigned int Y() const { return Y_; }
+    bool frameValid() const { return frameValid_; }
+    bool dataValid() const { return dataValid_; }
+    std::pair<int, int> cmsswIndex() const { return index_cmssw_; }
```
return const reference?

jbsauvan commented on L1Trigger/L1THGCal/plugins/backend/HGCalBackendLayer2Processor3DClustering_SA.cc:
```diff
@@ -90,19 +102,18 @@ class HGCalBackendLayer2Processor3DClusteringSA : public HGCalBackendLayer2Proce
-      }
-
-      for (const auto& collcluster : collCluster3D_perFPGA_sorted) {
+      for (const auto& collcluster : collCluster3D_perFPGA) {  //collCluster3D_perFPGA_sorted) {
```
Commented  code to be removed

jbsauvan commented on L1Trigger/L1THGCal/plugins/backend/HGCalHistoClusteringWrapper.cc:
```diff
@@ -21,166 +23,331 @@ class HGCalHistoClusteringWrapper : public HGCalHistoClusteringWrapperBase {
+      // But some of the TCs in the S1 emulation fall outside of the 60 degree region
+      // For now, assign these TCs to the 60 degree sector that the S2 emulation is expecting them to be in.
+      unsigned tcSector60 = iSector60;
+      unsigned int minSectorPhi = iSector60 * 648;
+      unsigned int maxSectorPhi = (iSector60 + 1) * 648;
```
Magic number `648` should be a named constant

EmyrClement replied:
Done

jbsauvan commented on L1Trigger/L1THGCal/plugins/backend/HGCalHistoClusteringWrapper.cc:
```diff
@@ -21,166 +23,331 @@ class HGCalHistoClusteringWrapper : public HGCalHistoClusteringWrapperBase {
+
+  // Distribute to links
+  clusters_SA.clear();
+  clusters_SA.resize(theConfiguration_.maxClustersPerLink() + 2);  // Two empty frames at start of packet
```
Instead of writing a comment, would be better to use a named constant for `2`, e.g.
```cpp
const unsigned empty_frames = 2;
clusters_SA.resize(theConfiguration_.maxClustersPerLink() + empty_frames);
```

jbsauvan commented on L1Trigger/L1THGCal/plugins/backend/HGCalHistoClusteringWrapper.cc:
```diff
@@ -21,166 +23,331 @@ class HGCalHistoClusteringWrapper : public HGCalHistoClusteringWrapperBase {
+    unsigned iCluster = 0;
+    for (auto& cluster : sector60) {
+      // Leave first two frames empty
+      unsigned frame = 2 + iCluster / nLinksPerSector60;
```
Magic number `2` to be replaced by named constant

jbsauvan commented on L1Trigger/L1THGCal/plugins/backend/HGCalHistoClusteringWrapper.cc:
```diff
@@ -21,166 +23,331 @@ class HGCalHistoClusteringWrapper : public HGCalHistoClusteringWrapperBase {
+      unsigned link = iCluster % nLinksPerSector60 + iSector60 * nLinksPerSector60;
+      if (frame >= theConfiguration_.maxClustersPerLink() + 2)
+        break;
+      clusters_SA[frame][link] = move(cluster);
```
`std::move` to make clear it is part of `std`

jbsauvan commented on L1Trigger/L1THGCal/plugins/backend/HGCalHistoClusteringWrapper.cc:
```diff
@@ -21,166 +23,331 @@ class HGCalHistoClusteringWrapper : public HGCalHistoClusteringWrapperBase {
+      // Leave first two frames empty
+      unsigned frame = 2 + iCluster / nLinksPerSector60;
+      unsigned link = iCluster % nLinksPerSector60 + iSector60 * nLinksPerSector60;
+      if (frame >= theConfiguration_.maxClustersPerLink() + 2)
```
Magic number `2`

jbsauvan commented on L1Trigger/L1THGCal/plugins/backend/HGCalHistoClusteringWrapper.cc:
```diff
@@ -21,166 +23,331 @@ class HGCalHistoClusteringWrapper : public HGCalHistoClusteringWrapperBase {
-        multiclusters.at(imulticluster).addConstituent(clustersPtrs.at(constituent.index_cmssw()), 1.);
-      }
-      firstConstituent = false;
+    if (pt < 0.5)
```
If cut value not taken from config, should be a named constant

EmyrClement replied:
Done, added to config.

jbsauvan commented on L1Trigger/L1THGCal/plugins/backend/HGCalHistoClusteringWrapper.cc:
```diff
@@ -21,166 +23,331 @@ class HGCalHistoClusteringWrapper : public HGCalHistoClusteringWrapperBase {
+    eta *= theConfiguration_.zSide();
+
+    auto sector = theConfiguration_.sector();
+    if (sector == 1) {
+      phi += (2. * M_PI / 3.);
+    } else if (sector == 2) {
+      phi = phi + (4. * M_PI / 3.);
     }
```
This seems to be repeated. Might be good to implement the rotation in some function

EmyrClement replied:
Moved to a function.

jbsauvan commented on L1Trigger/L1THGCal/plugins/backend/HGCalHistoClusteringWrapper.cc:
```diff
@@ -21,166 +23,331 @@ class HGCalHistoClusteringWrapper : public HGCalHistoClusteringWrapperBase {
+    multicluster.setP4(clusterP4);
+    std::cout << "Got a cluster : " << cluster->e() << " " << cluster->constituents().size() << " " << multicluster.pt()
+              << " " << multicluster.eta() << " " << multicluster.phi() << " " << rOverZ << " " << phi << " " << eta
+              << std::endl;
```
`cout` to be removed.

jbsauvan commented on L1Trigger/L1THGCal/plugins/backend/HGCalHistoClusteringWrapper.cc:
```diff
@@ -21,166 +23,331 @@ class HGCalHistoClusteringWrapper : public HGCalHistoClusteringWrapperBase {
+      multicluster.addConstituent(tc_cmssw, false, 0.);
+    }
+
+    double emIntFraction = 1.0 * cluster->e_em() / cluster->e();
```
Why is the `1.0*` needed?
If this is for conversion into float, then it would be better to have an explicit conversion

EmyrClement replied:
Using explicit conversion to float.

jbsauvan commented on L1Trigger/L1THGCal/plugins/backend/HGCalHistoClusteringWrapper.cc:
```diff
@@ -21,166 +23,331 @@ class HGCalHistoClusteringWrapper : public HGCalHistoClusteringWrapperBase {
+    multicluster.saveEnergyInterpretation(l1t::HGCalMulticluster::EnergyInterpretation::EM,
+                                          emIntFraction * multicluster.energy());
+
+    double emCoreIntFraction = 1.0 * cluster->e_em_core() / cluster->e();
```
ditto

jbsauvan commented on L1Trigger/L1THGCal/plugins/backend/HGCalHistoClusteringWrapper.cc:
```diff
@@ -21,166 +23,331 @@ class HGCalHistoClusteringWrapper : public HGCalHistoClusteringWrapperBase {
 
-  for (const auto& multicluster : multiclusters) {
-    multiclustersBXCollection.push_back(0, multicluster);
+    double emHEarlyIntFraction = 1.0 * cluster->e_h_early() / cluster->e();
```
ditto.

jbsauvan commented on L1Trigger/L1THGCal/plugins/backend/HGCalHistoClusteringWrapper.cc:
```diff
@@ -21,166 +23,331 @@ class HGCalHistoClusteringWrapper : public HGCalHistoClusteringWrapperBase {
-  for (const auto& multicluster : multiclusters) {
-    multiclustersBXCollection.push_back(0, multicluster);
+    double emHEarlyIntFraction = 1.0 * cluster->e_h_early() / cluster->e();
+    multicluster.saveEnergyInterpretation(l1t::HGCalMulticluster::EnergyInterpretation::H_EARLY,
```
energy interpretations are not really meant for this kind of energy values. But ok.

jbsauvan reviewed (commented):
Review in progress. Not finalized.

jbsauvan commented on L1Trigger/L1THGCal/python/hgcalBackendLayer2_fwClustering_cfi.py:
```diff
@@ -0,0 +1,132 @@
+          911 , 1020 , 1129 , 1238 , 1347 , 1456 , 1565 , 1674 , # CE-FH
+          1783 , 1892 , 2001 , 2110 , 2281 , 2452 , 2623 , 2794 , 2965 , 3136 , 3307 , 3478 , 3649 , 3820 , # CE-BH
+          0 , 0 , 0 , 0 , 0 , 0 , 0 , 0 , 0 , 0 , 0 , 0 , 0 ),
+    triggerLayers = cms.vuint32(0 , # No zero layer
+                  1 , 0 , 2 , 0 , 3 , 0 , # CE-E (early)
+                  4 , 0 , 5 , 0 , 6 , 0 , 7 , 0 , 8 , 0 , # CE-E (core)
+                  9 , 0 , 10 , 0 , 11 , 0 , 12 , 0 , 13 , 0 , 14 , 0 , # CE-E (back)
+                  15 , 16 , 17 , 18 , # CE-H (early)
+                  19 , 20 , 21 , 22 , 23 , 24 , 25 , 26 , 27 , 28 , 29 , 30 , 31 , 32 , 33 , 34 , 35 , 36 , 0 , 0 , # CE-H (back)
+                  0 , 0 , 0 , 0 , 0 , 0 , 0 , 0 , 0 , 0 , 0  ), # Padding
```
This is something that is somehow available through the trigger geometry class, but maybe not in this format.
Providing these numbers manually might be the source of typos and errors. Is it really needed to provide this manually through the config?

EmyrClement replied:
I think it would be better to derive this from the trigger geometry class, but as you say I don't think the info is accessible in this format (mapping from a layer to a trigger layer).  For now I could modify the trigger geometry class to provide this info (either here or in another PR?).  Longer term I would need to understand if the firmware (and emulator) really needs this info in this particular format or not.

jbsauvan replied:
That would be great if you can modify the geometry class. 
Rather in an other PR.
Looking quickly into the geometry class, I think adding a getter for `trigger_layers_` could do more or less the job.

EmyrClement replied:
Done

jbsauvan commented on L1Trigger/L1THGCal/python/hgcalBackendLayer2_fwClustering_cfi.py:
```diff
@@ -0,0 +1,132 @@
+                         0 , 0 , 0 , 0 , 0 , 0 , # CE-E (back)
+                         1 , 1 , 1 , 1 , # CE-H (early)
+                         0 , 0 , 0 , 0 , 0 , 0 , 0 , 0 , 0 , 0 , 0 , 0 , 0 , 0 , 0 , 0 , 0 , 0), # CE-H 
+    correction=cms.uint32(131071), # 0b011111111111111111
+    saturation=cms.uint32(524287) # (2 ** 19) - 1
```
General comment on how to write parameter values: wouldn't it be better to provide values in a more human-readable way, e.g `(2**19)-1` instead of `524287`?

EmyrClement replied:
Agreed, implemented.

jbsauvan commented on L1Trigger/L1THGCal/src/backend_emulator/DistServer.cc:
Since the access of `inputs_` is internal, could use `inputs_` instead of `inputs()`.
This way also the method `inputs()` could return a const ref (see one of my previous comments), and avoid having a getter function that allows to modify the internal state of the object.

EmyrClement replied:
Done

jbsauvan commented on L1Trigger/L1THGCal/src/backend_emulator/DistServer.cc:
```diff
@@ -0,0 +1,63 @@
+using namespace l1thgcfirmware;
+
+DistServer::DistServer(unsigned int nInputs, unsigned int nOutputs, unsigned int nInterleaving)
+    : nInputs_(nInputs), nOutputs_(nOutputs), nInterleaving_(nInterleaving), inputs_(nInputs_) {
```
`nInputs_` is redundant with `inputs_` (from which we can retrieve the size).
Is it ensured that `nInputs_` and `inputs_` stay in sync?

EmyrClement replied:
Removed nInputs_, just use inputs_.size() when needed.

jbsauvan commented on L1Trigger/L1THGCal/src/backend_emulator/DistServer.cc:
Same comment as before regarding the internal access to class attributes. Could use directly `addr_` and make the getter function return a const ref

EmyrClement replied:
Use `addr_` directly, removed getter function as no longer used.

jbsauvan commented on L1Trigger/L1THGCal/src/backend_emulator/DistServer.cc:
Same as before regarding `inputs()` and `inputs_`

jbsauvan commented on L1Trigger/L1THGCal/src/backend_emulator/DistServer.cc:
`std::rotate` to make it clear that this is a function from the standard library

jbsauvan commented on L1Trigger/L1THGCal/src/backend_emulator/DistServer.cc:
It seems that `toRotate` can be a const reference

EmyrClement replied:
I don't think this will work, as `toRotate` is modified on the next line.

jbsauvan replied:
right

jbsauvan commented on L1Trigger/L1THGCal/src/backend_emulator/HGCalCluster_SA.cc:
```diff
@@ -0,0 +1,69 @@
+  this->set_sat_tc(this->sat_tc() | c.sat_tc());
+  this->set_shapeq(this->shapeq() | c.shapeq());
+
+  if (w_ <= 52438 && original.shapeq() == 1 &&
```
Magic numbers to be changed to named constants

jbsauvan commented on L1Trigger/L1THGCal/src/backend_emulator/HGCalHistoClusterProperties_SA.cc:
```diff
@@ -0,0 +1,210 @@
+
+  for (const auto& flag : readoutFlags) {
+    auto accumulator = make_unique<HGCalCluster>(0, 0, true, true);
+    flag->setClock(flag->clock() + 23);  // Magic numbers - latency of which particular step?
```
Magic numbers to be turned into a named constant

jbsauvan commented on L1Trigger/L1THGCal/src/backend_emulator/HGCalHistoClusterProperties_SA.cc:
```diff
@@ -0,0 +1,210 @@
+    accumulator->setDataValid(true);
+
+    if (sums.find(flag->clock()) == sums.end()) {
+      auto sum = make_unique<HGCalCluster>(flag->clock() + 7,  // Magic numbers - latency of which particular step?
```
magic numbers to be turned into a named constant

jbsauvan commented on L1Trigger/L1THGCal/src/backend_emulator/HGCalHistoClusterProperties_SA.cc:
```diff
@@ -0,0 +1,210 @@
+  //ID++;
+  //nTCs += c->n_tc();
+  //}
+  // std::cout << nTCs << std::endl;
```
Commented code to be removed

jbsauvan commented on L1Trigger/L1THGCal/src/backend_emulator/HGCalHistoClusterProperties_SA.cc:
```diff
@@ -0,0 +1,210 @@
+// Calculates properties of clusters from accumulated quantities
+void HGCalHistoClusterProperties::clusterProperties(HGCalClusterSAPtrCollection& clusterSums) const {
+  unsigned int nTCs = 0;
+  //  std::cout << "Cluster Prop" << std::endl;
```
commented code to be removed

jbsauvan commented on L1Trigger/L1THGCal/src/backend_emulator/HGCalHistoClusterProperties_SA.cc:
```diff
@@ -0,0 +1,210 @@
+void HGCalHistoClusterProperties::clusterProperties(HGCalClusterSAPtrCollection& clusterSums) const {
+  unsigned int nTCs = 0;
+  //  std::cout << "Cluster Prop" << std::endl;
+  for (const auto& c : clusterSums) {
```
Not clear why `const` is used here since `clusterSums` is non-const and the clusters are modified inside the loop.

EmyrClement replied:
Removed `const`

jbsauvan commented on L1Trigger/L1THGCal/src/backend_emulator/HGCalHistoClusterProperties_SA.cc:
```diff
@@ -0,0 +1,210 @@
+    c->set_E_H_early_over_E_Quotient(E_H_early_over_E.first);
+    c->set_E_H_early_over_E_Fraction(E_H_early_over_E.second);
+
+    // std::cout << c->clock() << " " << c->index() << " " << c->n_tc() << " " << c->e() << " " << c->e_em() << " " << c->e_em_core() << " " << c->e_h_early() << " N "<< c->Sigma_E_Quotient()<< " " <<c->Sigma_E_Fraction() << " " << c->Mean_z_Quotient() << " " << c->Mean_z_Fraction() << " " << c->Mean_phi_Quotient() << " "<< c->Mean_phi_Fraction() << " " << c->Mean_eta_Quotient() << " " << c->Mean ...
```
Commented code to be removed

jbsauvan commented on L1Trigger/L1THGCal/src/backend_emulator/HGCalHistoClusterProperties_SA.cc:
```diff
@@ -0,0 +1,210 @@
+    // std::cout << c->clock() << " " << c->index() << " " << c->n_tc() << " " << c->e() << " " << c->e_em() << " " << c->e_em_core() << " " << c->e_h_early() << " N "<< c->Sigma_E_Quotient()<< " " <<c->Sigma_E_Fraction() << " " << c->Mean_z_Quotient() << " " << c->Mean_z_Fraction() << " " << c->Mean_phi_Quotient() << " "<< c->Mean_phi_Fraction() << " " << c->Mean_eta_Quotient() << " " << c->Mean ...
+    nTCs += c->n_tc();
+  }
+  //  std::cout << nTCs << std::endl;
```
Commented code to be removed

jbsauvan commented on L1Trigger/L1THGCal/src/backend_emulator/HGCalHistoClusterProperties_SA.cc:
```diff
@@ -0,0 +1,210 @@
+  unsigned long int N = N_TC_W * Sum_W2 - pow(Sum_W, 2);
+  unsigned long int D = pow(N_TC_W, 2);
+  double intpart;
+  double frac = modf(sqrt(N / D), &intpart) * pow(2, 1);
```
D!=0 to be checked

EmyrClement replied:
Done

jbsauvan commented on L1Trigger/L1THGCal/src/backend_emulator/HGCalHistoClusterProperties_SA.cc:
```diff
@@ -0,0 +1,210 @@
+std::pair<unsigned int, unsigned int> HGCalHistoClusterProperties::mean_coordinate(unsigned int Sum_Wc,
+                                                                                   unsigned int Sum_W) const {
+  double intpart;
+  double frac = modf((double)Sum_Wc / Sum_W, &intpart) * pow(2, 2);
```
- Sum_W!=0 to be checked 
- `pow(2,2)` to be replaced with `4` (and using a named constant)

EmyrClement replied:
Done, also for similar comments below.

jbsauvan commented on L1Trigger/L1THGCal/src/backend_emulator/HGCalHistoClusterProperties_SA.cc:
```diff
@@ -0,0 +1,210 @@
+  unsigned long int N = Sum_W * Sum_Wc2 - pow(Sum_Wc, 2);
+  unsigned long int D = pow(Sum_W, 2);
+  double intpart;
+  double frac = modf((double)sqrt(N / D), &intpart) * pow(2, 1);
```
- D!=0 to be checked
- `pow(2,1)` to be replaced with `2` (using a named constant)

jbsauvan commented on L1Trigger/L1THGCal/src/backend_emulator/HGCalHistoClusterProperties_SA.cc:
```diff
@@ -0,0 +1,210 @@
+    return {0, 0};
+  } else {
+    double intpart;
+    double frac = modf((double)E_N / E_D, &intpart) * pow(2, 8);
```
Same comments regarding E_D!=0 and `pow(2,8)`

jbsauvan commented on L1Trigger/L1THGCal/src/backend_emulator/HGCalHistoClusterProperties_SA.cc:
```diff
@@ -0,0 +1,210 @@
+  int lastLayer = 0;
+  std::vector<int> layerBits_array;
+
+  for (int idx = 0; idx < 36; idx++) {
+    if ((layerBits & (1L << (35 - idx))) >= 1L) {
```
Magic numbers to be turned into named constants

EmyrClement replied:
Done, but will try to extract from config/CMSSW as this is just the number of trigger layers.

EmyrClement replied:
Done

jbsauvan commented on L1Trigger/L1THGCal/src/backend_emulator/HGCalHistoClusterProperties_SA.cc:
```diff
@@ -0,0 +1,210 @@
+    }
+  }
+  int showerLen = lastLayer - firstLayer + 1;
+  int coreShowerLen = 36;
```
Magic number to be turned into named constant

jbsauvan commented on L1Trigger/L1THGCal/src/backend_emulator/HGCalHistoClusterProperties_SA.cc:
```diff
@@ -0,0 +1,210 @@
+    coreShowerLen = *std::max_element(layerBits_array.begin(), layerBits_array.end());
+  }
+
+  std::vector<int> output = {firstLayer, lastLayer, showerLen, coreShowerLen};
+
+  return output;
```
Could return directly the `{...}`

EmyrClement replied:
Done

jbsauvan commented on L1Trigger/L1THGCal/src/backend_emulator/HGCalHistoClusteringConfig_SA.cc:
```diff
@@ -0,0 +1,145 @@
+      layerWeights_E_EM_(),
+      layerWeights_E_EM_core_(),
+      layerWeights_E_H_early_(),
+      correction_(),  // 0b011111111111111111
+      saturation_()   // (2 ** 19) - 1
```
The comments are misleading as these are not the initialization values

EmyrClement replied:
Comments removed.

jbsauvan commented on L1Trigger/L1THGCal/src/backend_emulator/HGCalHistoClusteringConfig_SA.cc:
```diff
@@ -0,0 +1,145 @@
+  // Add check that stepLatency is at least same size as latencies
+  // But not as cms.exception
+  for (unsigned int iStep = 0; iStep < latencies.size(); ++iStep) {
+    stepLatency_.at(Step(iStep)) = latencies.at(iStep);
```
I'm not sure to see the point of converting to the `Step` enum and then back to an int in `at()`.
And I guess in case `iStep` is out of range compared to the values defined in the enum, the behaviour is undefined.

EmyrClement replied:
`stepLatency_` is now a vector, so this comment no longer applies.

jbsauvan commented on L1Trigger/L1THGCal/src/backend_emulator/HGCalHistoClusteringConfig_SA.cc:
```diff
@@ -0,0 +1,145 @@
+  const unsigned int lTarget = int((maxBinsSmearing1D_ + 0.5) * lWidth0 - 0.5);
+  for (unsigned int iBin = 0; iBin < bins; ++iBin) {
+    unsigned int lCentre = lWidth0 + (height * iBin);
+    const unsigned int lBins = int(round(1.0 * lTarget / lCentre));
```
Better to remove `1.0` and have an explicit conversion to `float`

EmyrClement replied:
Done

jbsauvan commented on L1Trigger/L1THGCal/src/backend_emulator/HGCalHistoClusteringConfig_SA.cc:
```diff
@@ -0,0 +1,145 @@
+
+    lCentre *= lBins;
+
+    const unsigned int lRatio = int(round(1.0 * lTarget / lCentre * pow(2, nBitsAreaNormLUT_)));
```
- Better to remove `1.0` and have an explicit conversion
- `pow(2, nBitsAreaNormLUT_)` to be replaced by `0x1<<nBitsAreaNormLUT_`, which is more efficient

EmyrClement replied:
Done, same for next comment

jbsauvan commented on L1Trigger/L1THGCal/src/backend_emulator/HGCalHistoClusteringConfig_SA.cc:
```diff
@@ -0,0 +1,145 @@
+
+void ClusterAlgoConfig::initializeCosLUT() {
+  for (unsigned int iBin = 0; iBin < nBinsCosLUT_ + 1; ++iBin) {
+    unsigned int cosBin = round(pow(2, nBitsCosLUT_) * (1 - cos(iBin * M_PI / phiNValues_)));
```
`pow(2, nBitsCosLUT_) ` to be replaced with `0x1<<nBitsCosLUT_`

jbsauvan commented on L1Trigger/L1THGCal/src/backend_emulator/HGCalHistoClusteringConfig_SA.cc:
```diff
@@ -0,0 +1,145 @@
+  }
+}
+
+void ClusterAlgoConfig::printConfiguration() {
```
Should try to find a way to avoid `cout` in CMSSW. Maybe just returning a string, which is then given to the message logger in CMSSW and to standard cout in the standalone code.

EmyrClement replied:
This was mainly for debugging early on in development.  I am happy to delete it.  It may still be useful to dump something like this via the MessageLogger or to file, but not in this way.

jbsauvan commented on L1Trigger/L1THGCal/src/backend_emulator/HGCalHistoClusteringImpl_SA.cc:
```diff
@@ -0,0 +1,28 @@
+void HGCalHistoClusteringImplSA::runAlgorithm(const HGCalTriggerCellSAPtrCollections& inputs,
+                                              HGCalTriggerCellSAShrPtrCollection& clusteredTCs,
+                                              HGCalClusterSAPtrCollection& clusterSums) const {
+  // config_.printConfiguration();
```
Commented code to be removed

jbsauvan commented on L1Trigger/L1THGCal/src/backend_emulator/HGCalHistoClustering_SA.cc:
```diff
@@ -0,0 +1,330 @@
+  CentroidHelperPtrCollections fifos(config_.nFifos());
+  vector<unsigned int> clock(config_.cColumns(), config_.clusterizerMagicTime());
+  CentroidHelperShrPtrCollection latched(
+      config_.nFifos() + 1,
```
Magic number `1` to be turned into a named constant, e.g.
```cpp
const unsigned dummy_entries = 1;
```

jbsauvan commented on L1Trigger/L1THGCal/src/backend_emulator/HGCalHistoClustering_SA.cc:
```diff
@@ -0,0 +1,330 @@
+    }
+
+    CentroidHelperShrPtrCollection accepted(
+        config_.nFifos() + 2,
```
Magic number `2` to be changed into a named constant.
Besides the comment below mentions a value of `1`

EmyrClement replied:
Done, and clarified comment.  It's one dummy entry compared to `latched`, but two compared to `fifos`.

jbsauvan commented on L1Trigger/L1THGCal/src/backend_emulator/HGCalHistoClustering_SA.cc:
```diff
@@ -0,0 +1,330 @@
+        for (unsigned int iCol = a->column() - config_.nColumnsForClustering();
+             iCol < a->column() + config_.nColumnsForClustering() + 1;
+             ++iCol) {
+          clock[iCol] += 8;  // Magic numbers - latency of which particular step?
```
Magic number to be changed into named constant

jbsauvan commented on L1Trigger/L1THGCal/src/backend_emulator/HGCalHistoClustering_SA.cc:
```diff
@@ -0,0 +1,330 @@
+              unsigned int dR2 = dR * dR;
+              unsigned int cosTerm = (absDPhi > config_.nBinsCosLUT()) ? 2047 : config_.cosLUT(absDPhi);
+              dR2 += int(r1 * r2 / pow(2, 7)) * cosTerm /
+                     pow(2, 10);  // Magic numbers - number of bits involved in deltaR calculation?
```
`pow(2,7)` and `pow(2,10)` to be replaced with `128` and  `1024`, using named constants

jbsauvan commented on L1Trigger/L1THGCal/src/backend_emulator/HGCalHistoClustering_SA.cc:
```diff
@@ -0,0 +1,330 @@
+
+          CentroidHelperPtr readoutFlag = make_unique<CentroidHelper>(T - 2, iCol, true);
+          if (readoutFlag->clock() ==
+              config_.clusterizerMagicTime() + 14) {  // Magic numbers - latency of which particular step?
```
Magic number to be turned into a named constant

jbsauvan commented on L1Trigger/L1THGCal/src/backend_emulator/HGCalHistoClustering_SA.cc:
```diff
@@ -0,0 +1,330 @@
+    }
+  }
+
+  for (unsigned int i = 0; i < 1000;
```
Magic number to be turned into a named constant

jbsauvan commented on L1Trigger/L1THGCal/src/backend_emulator/HGCalHistoClustering_SA.cc:
```diff
@@ -0,0 +1,330 @@
+  // std::cout << "Number of readoutFlags : " << readoutFlagsOut.size() << std::endl;
+  // for ( const auto& f : readoutFlagsOut ) {
+  //   std::cout << f->clock() << " " << f->index() << " " << f->column() << " " << f->row() << " " << f->energy() << " " << f->X() << " " << f->Y() << " " << f->dataValid() << std::endl;
+  // }
```
commented code to be removed

jbsauvan commented on L1Trigger/L1THGCal/src/backend_emulator/HGCalHistoClustering_SA.cc:
```diff
@@ -0,0 +1,330 @@
+    // Cluster from single TC
+    // Does this ever happen?
+    // Removed from newer versions of firmware in any case
+    if (tc->deltaR2() >= 25000) {  // Magic numbers - but removed in newer versions of firmware, so leave for now
```
Magic number to be turned into a named constant

jbsauvan commented on L1Trigger/L1THGCal/src/backend_emulator/HGCalHistoClustering_SA.cc:
```diff
@@ -0,0 +1,330 @@
+      continue;
+    }
+
+    unsigned long int s_TC_W = (int(tc->energy() / 4) == 0) ? 1 : tc->energy() / 4;
```
Magic number `4` to be turned into a named constant

jbsauvan commented on L1Trigger/L1THGCal/src/backend_emulator/HGCalHistoClustering_SA.cc:
```diff
@@ -0,0 +1,330 @@
+
+    unsigned int triggerLayer = config_.triggerLayer(tc->layer());
+    unsigned int s_E_EM =
+        ((((unsigned long int)tc->energy() * config_.layerWeight_E_EM(triggerLayer)) + config_.correction()) >> 18);
```
Magic number `18` to be turned into a named constant

jbsauvan commented on L1Trigger/L1THGCal/src/backend_emulator/HGCalHistoClustering_SA.cc:
```diff
@@ -0,0 +1,330 @@
+      s_E_EM = config_.saturation();
+
+    unsigned int s_E_EM_core =
+        (((unsigned long int)tc->energy() * config_.layerWeight_E_EM_core(triggerLayer) + config_.correction()) >> 18);
```
Magic number `18` to be turned into a named constant

jbsauvan commented on L1Trigger/L1THGCal/src/backend_emulator/HGCalHistoClustering_SA.cc:
```diff
@@ -0,0 +1,330 @@
+    cluster->set_wphi2(s_TC_W * tc->phi() * tc->phi());
+    cluster->set_wroz2(s_TC_W * tc->rOverZ() * tc->rOverZ());
+
+    cluster->set_layerbits(cluster->layerbits() | (((unsigned long int)1) << (36 - triggerLayer)));
```
Magic number `36` to be turned into a named constant

jbsauvan commented on L1Trigger/L1THGCal/src/backend_emulator/HGCalHistoClustering_SA.cc:
```diff
@@ -0,0 +1,330 @@
+  // for ( const auto& pclus : protoClusters ) {
+
+  //     std::cout << pclus->clock() << " " << pclus->index() << " " << pclus->n_tc() << " " << pclus->e() << " " << pclus->e_em() << " " << pclus->e_em_core() << " " << pclus->e_h_early() << " " << pclus->w() << " " << pclus->n_tc_w() << " " << pclus->weta2() << " " << pclus->wphi2() << std::endl;
+  // }
```
Commented code to be removed

jbsauvan commented on L1Trigger/L1THGCal/src/backend_emulator/HGCalHistoSeeding_SA.cc:
```diff
@@ -0,0 +1,232 @@
+
+      if (width >= 2) {
+        if (int(col - offset - 1) >= 0) {
+          l2 = make_unique<HGCalHistogramCell>(lHistogram[binIndex - offset - 1] / 4);
```
Magic number `4` to be turned into a named constant

jbsauvan commented on L1Trigger/L1THGCal/src/backend_emulator/HGCalHistoSeeding_SA.cc:
```diff
@@ -0,0 +1,232 @@
+          l2 = make_unique<HGCalHistogramCell>(lHistogram[binIndex - offset - 1] / 4);
+        }
+        if (int(col + offset + 1) <= int(config_.cColumns() - 1)) {
+          r2 = make_unique<HGCalHistogramCell>(lHistogram[binIndex + offset + 1] / 4);
```
Magic number `4` to be turned into a named constant

jbsauvan commented on L1Trigger/L1THGCal/src/backend_emulator/HGCalHistoSeeding_SA.cc:
```diff
@@ -0,0 +1,232 @@
+      }
+
+      if (int(col - offset) >= 0) {
+        l1 = make_unique<HGCalHistogramCell>(lHistogram[binIndex - offset] / 2);
```
Magic number `2` to be turned into a named constant

jbsauvan commented on L1Trigger/L1THGCal/src/backend_emulator/HGCalHistoSeeding_SA.cc:
```diff
@@ -0,0 +1,232 @@
+        l1 = make_unique<HGCalHistogramCell>(lHistogram[binIndex - offset] / 2);
+      }
+      if (int(col + offset) <= int(config_.cColumns() - 1)) {
+        r1 = make_unique<HGCalHistogramCell>(lHistogram[binIndex + offset] / 2);
```
Magic number `2` to be turned into a named constant

jbsauvan commented on L1Trigger/L1THGCal/src/backend_emulator/HGCalHistoSeeding_SA.cc:
```diff
@@ -0,0 +1,232 @@
+      *hc += ((*l2 + *l1) / scale + (*r2 + *r1) / scale);
+      scale *= 4;
+      width -= 2;
+      offset += 2;
```
Magic numbers to be turned into named constants

jbsauvan commented on L1Trigger/L1THGCal/src/backend_emulator/HGCalHistoSeeding_SA.cc:
```diff
@@ -0,0 +1,232 @@
+  for (unsigned int iBin = 0; iBin < histogram.size(); ++iBin) {
+    HGCalHistogramCell& hc = *histogram.at(iBin);
+    hc.addLatency(stepLatency);
+    hc.setS(int(1.0 * hc.S() * config_.areaNormalization(hc.sortKey()) / pow(2, config_.nBitsAreaNormLUT() + 1)));
```
- Multiplication by `1.0` to be replaced by explicit conversion to float?
- `pow(2, config_.nBitsAreaNormLUT()` to be replaced by `0x1<<config_.nBitsAreaNormLUT()`, more efficient to compute

EmyrClement replied:
Done both

jbsauvan commented on L1Trigger/L1THGCal/src/backend_emulator/HGCalHistoSeeding_SA.cc:
```diff
@@ -0,0 +1,232 @@
+    const int row = hc->sortKey();
+    const unsigned int binIndex = config_.cColumns() * row + col;
+    if (row - 1 >= 0) {
+      *hc += (lHistogram[binIndex - config_.cColumns()] / 2);
```
Magic number `2` to be turned into named constant

jbsauvan commented on L1Trigger/L1THGCal/src/backend_emulator/HGCalHistoSeeding_SA.cc:
```diff
@@ -0,0 +1,232 @@
+      *hc += (lHistogram[binIndex - config_.cColumns()] / 2);
+    }
+    if (row + 1 <= int(config_.cRows() - 1)) {
+      *hc += (lHistogram[binIndex + config_.cColumns()] / 2);
```
Magic number `2` to be turned into named constant

jbsauvan commented on L1Trigger/L1THGCal/src/backend_emulator/HGCalHistoSeeding_SA.cc:
```diff
@@ -0,0 +1,232 @@
+// Temporary simulation of local maxima finder
+// Not an emulation of any firmware
+void HGCalHistoSeeding::localMaximaFinder(HGCalHistogramCellSAPtrCollection& histogram) const {
+  // const std::vector<unsigned> maximaWidths{ 6, 5, 5, 5, 4, 4, 4, 4, 3, 3, 3, 3, 3, 3, 3, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1 }; // Padded this with 4*1 at the end.  Firmware this was based on has 40 entries, but there are 44 bins.
```
Commented code to be removed

jbsauvan commented on L1Trigger/L1THGCal/src/backend_emulator/HGCalHistoSeeding_SA.cc:
```diff
@@ -0,0 +1,232 @@
+  for (auto& hc : histogram) {
+    hc->addLatency(stepLatency);
+    if (hc->N() > 0) {
+      unsigned int inv_N = int(round(1.0 * 0x1FFFF / hc->N()));
```
- Magic number `0x1FFFF` to be turned into named constant
- Multiplication by `1.0` to be replaced by an explicit conversion to float?

jbsauvan commented on L1Trigger/L1THGCal/src/backend_emulator/HGCalHistoSeeding_SA.cc:
```diff
@@ -0,0 +1,232 @@
+    hc->addLatency(stepLatency);
+    if (hc->N() > 0) {
+      unsigned int inv_N = int(round(1.0 * 0x1FFFF / hc->N()));
+      hc->setX((hc->X() * inv_N) >> 17);
+      hc->setY((hc->Y() * inv_N) >> 17);
```
Magic number `17` to be turned into named constant

jbsauvan commented on L1Trigger/L1THGCal/src/backend_emulator/HGCalHistoSeeding_SA.cc:
```diff
@@ -0,0 +1,232 @@
+}
+
+// Useful debugging function for printing histogram contents
+void HGCalHistoSeeding::printHistogram(const HGCalHistogramCellSAPtrCollection& histogram) const {
```
Same as one of the previous comments.
Should avoid `cout` in CMSSW, so could create a string and the actual printing is done outside, with message logger in CMSSW.

EmyrClement replied:
Using `MessageLogger` for printing histogram.

jbsauvan commented on L1Trigger/L1THGCal/src/backend_emulator/HGCalHistogramCell_SA.cc:
```diff
@@ -0,0 +1,28 @@
+const HGCalHistogramCell HGCalHistogramCell::operator/(const unsigned int factor) const {
+  HGCalHistogramCell hc(*this);
+
+  hc.S_ /= factor;
```
check if factor!=0

jbsauvan commented on L1Trigger/L1THGCalUtilities/plugins/ntuples/HGCalTriggerNtupleHGCMulticlustersSA.cc:
```diff
@@ -0,0 +1,191 @@
+#include "L1Trigger/L1THGCal/interface/HGCalTriggerGeometryBase.h"
+#include "L1Trigger/L1THGCalUtilities/interface/HGCalTriggerNtupleBase.h"
+
+class HGCalTriggerNtupleHGCMulticlustersSA : public HGCalTriggerNtupleBase {
```
Instead of duplicating the multi-cluster ntuple, might be better to add variables in the existing one, and eventually have a flag specifying which sets of variables we want to store

EmyrClement replied:
I've modified the existing multicluster ntuple to include the additional hw based variables, and a python config option specifying whether to store these or not.

jbsauvan reviewed (commented):
Second set of comments. Review finalized.
Thanks a lot @EmyrClement for the updates and the cleaning of the code

EmyrClement commented:
@jbsauvan I believe all comments are addressed, other than [this one](https://github.com/hgc-tpg/cmssw/pull/5#pullrequestreview-1134009706).

EmyrClement commented:
I addressed the last comment (and re-ran `scram b code-checks` and `scram b code-format`).

jbsauvan commented on L1Trigger/L1THGCal/interface/backend_emulator/HGCalHistoClusteringConfig_SA.h:
```diff
@@ -154,54 +157,55 @@ namespace l1thgcfirmware {
-        triggerLayers_.push_back(triggerLayer);
-    }
-    std::vector<unsigned int> triggerLayers() const { return triggerLayers_; }
+    void setTriggerLayers(const std::vector<unsigned int> triggerLayers) { triggerLayers_ = triggerLayers; }
```
Could use const reference to pass `triggerLayers`

EmyrClement replied:
Done

jbsauvan commented on L1Trigger/L1THGCal/interface/HGCalTriggerGeometryBase.h:
```diff
@@ -82,6 +82,7 @@ class HGCalTriggerGeometryBase {
   virtual bool disconnectedModule(const unsigned module_id) const = 0;
   virtual unsigned lastTriggerLayer() const = 0;
   virtual unsigned triggerLayer(const unsigned id) const = 0;
+  virtual std::vector<unsigned> triggerLayers() const = 0;
```
Better to return a const reference

jbsauvan commented on L1Trigger/L1THGCal/interface/backend_emulator/HGCalHistoSeeding_SA.h:
```diff
@@ -4,12 +4,13 @@
 #include "L1Trigger/L1THGCal/interface/backend_emulator/HGCalTriggerCell_SA.h"
 #include "L1Trigger/L1THGCal/interface/backend_emulator/HGCalHistogramCell_SA.h"
 #include "L1Trigger/L1THGCal/interface/backend_emulator/HGCalHistoClusteringConfig_SA.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
```
This include won't be available when running the emulator standalone. Calls to MessageLogger should be done outside the emulator (e.g. in the wrapper).

EmyrClement replied:
I've removed the `printHistogram` function, it was only used for debugging.

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp2.cc:
```diff
@@ -59,6 +59,7 @@ class HGCalTriggerGeometryV9Imp2 : public HGCalTriggerGeometryBase {
   bool disconnectedModule(const unsigned) const final;
   unsigned lastTriggerLayer() const final { return last_trigger_layer_; }
   unsigned triggerLayer(const unsigned) const final;
+  std::vector<unsigned> triggerLayers() const final { return trigger_layers_; }
```
Better to return a const reference

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp3.cc:
```diff
@@ -62,6 +62,7 @@ class HGCalTriggerGeometryV9Imp3 : public HGCalTriggerGeometryBase {
   bool disconnectedModule(const unsigned) const final;
   unsigned lastTriggerLayer() const final { return last_trigger_layer_; }
   unsigned triggerLayer(const unsigned) const final;
+  std::vector<unsigned> triggerLayers() const final { return trigger_layers_; }
```
Better to return a const reference

jbsauvan commented on L1Trigger/L1THGCal/src/backend_emulator/HGCalCluster_SA.cc:
```diff
@@ -28,8 +28,8 @@ const HGCalCluster& HGCalCluster::operator+=(HGCalCluster& c) {
 
-  if (w_ <= 52438 && original.shapeq() == 1 &&
-      c.shapeq() == 1) {  // Magic numbers - possibly removed in newer versions of firmware?
+  const unsigned clusterWeightSat80 = (pow(2, 16) - 1) * 0.8;  // 52428
```
Could use `1<<16` instead of `pow(2,16)` (cheaper to compute)

jbsauvan commented on L1Trigger/L1THGCal/src/backend_emulator/HGCalHistoSeeding_SA.cc:
```diff
@@ -212,21 +218,27 @@ void HGCalHistoSeeding::calculateAveragePosition(HGCalHistogramCellSAPtrCollecti
-    std::cout << std::endl;
+    histogramString += "\n";
   }
+  edm::LogInfo("HGCalHistoSeeding") << "Histogram :";
+  edm::LogInfo("HGCalHistoSeeding") << histogramString;
```
Since MessageLogger won't be available in the standalone emulator, this should be outside of the emulator code

EmyrClement replied:
Removed function.

jbsauvan commented on L1Trigger/L1THGCalUtilities/python/hgcalTriggerNtuples_cfi.py:
```diff
@@ -92,8 +92,9 @@
     EGIdentification = egamma_identification_histomax.clone(),
-    FillLayerInfo = cms.bool(False),
-    FillInterpretationInfo = cms.bool(True)
+    FillLayerInfo = cms.bool(True),
```
`FillLayerInfo` should be kept as `False`, at least in this PR (since this change is not related to the emulator)

EmyrClement replied:
Apologies, this slipped through to a commit by accident, fixed.

jbsauvan replied:
Seems that it is still set to `True`

jbsauvan commented on L1Trigger/L1THGCalUtilities/python/stage2FileWriter_cff.py:
```diff
@@ -0,0 +1,7 @@
+import FWCore.ParameterSet.Config as cms
+
+Stage2FileWriter = cms.EDAnalyzer('Stage2FileWriter',
```
Is this implemented?

EmyrClement replied:
Apologies, slipped through after switching branches.  It's for the pattern file generation for the interface tests.  Removed (and for next comment).

jbsauvan commented on L1Trigger/L1THGCalUtilities/python/l1tInputFileReader_cff.py:
```diff
@@ -0,0 +1,6 @@
+import FWCore.ParameterSet.Config as cms
+
+L1TInputFileReader = cms.EDProducer('L1TInputFileReader',
```
Is this implemented?

jbsauvan reviewed (commented):
Thanks @EmyrClement 
Here is the last set of tiny comments.

jbsauvan commented:
Thanks @EmyrClement for the updates
Could you rebase on top of the latest dev branch (`hgc-tpg-devel-CMSSW_12_5_0_pre4`) (after the remaining change on the cluster layer energies filling)?

EmyrClement commented:
@jbsauvan I completed the rebase to `hgc-tpg-devel-CMSSW_12_5_0_pre4` - what's the preferred way of updating this PR?  Force push or a new PR for the rebased branch?

jbsauvan commented:
Thanks @EmyrClement 
You can force push and keep the same PR

EmyrClement commented:
@jbsauvan I pushed the rebase.  I noticed one [commit](https://github.com/hgc-tpg/cmssw/pull/5/commits/dee9332fd52387563740a81b3b9a4d1790e5deec) from Samuel still made it in to this PR after the rebase.  Should I remove it?

jbsauvan commented:
Yes, you can remove this commit. These data files are now in `cms-data`

EmyrClement commented:
Done
